lessons/lesson1.md

```diff
@@ -1,5 +1,167 @@
+/* this is a sample action sent a message to slack */
+var request = require('request');
+
+/* default slackwebhook and channel add yours here and replace the TODO below */
+var slackWebhook = "Your webhook";
+var slackChannel = "your channel";
+
+async function main (params) {
+  
+  /* print event detail */
+  console.log('in main + event detail: ', params.event);
+
+  var returnObject = {
+    statusCode: 200,
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: ""
+  };
+
+  /* handle the challenge */
+  if (params.challenge) {
+
+    console.log('Returning challenge: ' + params.challenge);
+
+    returnObject.body = new Buffer(JSON.stringify({
+      "challenge": params.challenge
+    })).toString('base64');
+
+    return returnObject;
+
+  } else {
+
+    /* we need it to run asynchronously, so we are returning a Promise */
+    return new Promise(function (resolve, reject) {
+
+      var slackMessage = " Event received: " + JSON.stringify(params);
+
+      var payload = {
+        "channel": slackChannel,
+        "username": "incoming-webhook",
+        "text": slackMessage,
+        "mrkdwn": true,
+      };
+
+      var options = {
+        method: 'POST',
+        url: slackWebhook,
+        headers:
+            { 'Content-type': 'application/json' },
+        body: JSON.stringify(payload)
+      };
+
+      request(options, function (error, response, body) {
+        if (error) {
+
+          console.log("ERROR: fail to post " + response);
+
+          reject(error);
+
+        } else {
+
+          console.log ("SUCCESS: posted to slack " + slackMessage);
+
+          returnObject.body = new Buffer(JSON.stringify({
+            "slackMessage": slackMessage
+          })).toString('base64');
+
+          resolve(returnObject);
+        }
+
+      });
+
+    });
+
+  }
+}
+
+exports.main = main
+```
+
+### Event Runtime Integration 
+
+- With all above setup, now you get your `providerId`, `eventCode`, you can go back to your firefly App trying to invoke a custom event like below: 
+![invoke-event](assets/event-invoke-4.png)
+
+- You should see this runtime action created in the `user defined actions` 
+![user-define-action](assets/user-define-action-5.png)
+
+- User now adds the event api to the project to setup the event registration
+![add-event](assets/add-event-6.png)
+
+- Adding from our custom event provider we just registered `eventruntime` (you should be able to see your register event in this list)
+![add-event](assets/add-event-7.png)
+
+- Subscribing to the "eventrt" event type
+![add-event](assets/add-event-8.png)
+
+- Generate the JWT service account credentials key pair
+![add-event](assets/add-event-9.png)
+
+- On the registration details page provide name and select the runtime user action created to setup event registration, select the user action from the dropdown of Runtime Actions, as we create the event consumer using generic/index.js from `generic` template, so we will choose this one
```

Comment:
I forgot one important point. An action used as event consumer does not need to be `web: yes`, and doesn't need `require-adobe-auth: true`. We should instruct users to remove them to assure app security.

Reply:
fixed